### ISSUE TITLE
Add a new private kubernetes cluster in the new sponsored azure subscription

### ISSUE THREAD

**Issue by smerle33** [jenkins-infra/helpdesk#3923]
### Service(s)

Azure, infra.ci.jenkins.io

### Summary

as per https://github.com/jenkins-infra/helpdesk/issues/3918#issuecomment-1911743537

- "Using subscription credits instead of CDF money" following https://github.com/jenkins-infra/helpdesk/issues/3818 and https://github.com/jenkins-infra/helpdesk/issues/3913
- "Continue work about infra.ci on arm64" as part of https://github.com/jenkins-infra/helpdesk/issues/3823

**=> Let's scope the initial implementation to only infra.ci.jenkins.io agents, and only 1 "non system" nodepool of type linux/arm64 so we can start switching workloads out of privatek8s.**

add a new AKS cluster only for infra.ci.jenkins.io and release.ci.jenkins.io Kubernetes agents

- [x] Create a new AKS cluster privatek8s-sponsored
  - [x] Managed in jenkins-infra/azure with Terraform see https://github.com/jenkins-infra/helpdesk/issues/3923#issuecomment-2147128098
  - [x] Same settings as the current [privatek8s] (https://github.com/jenkins-infra/azure/blob/0cf6c87851af116dc6b7ec955683753db8b47ce5/privatek8s.tf#L24) but with only nodepools for agents
  - [x] set network to manage bridge between the current privatek8s cluster and this one
- [x]   add this cluster to kubernetes-management
  - [x] Creates a service-account with token + insert it in infra.ci secrets
  - [x] Add initial "foundation" services (~~private + public ingress, cert-manager,~~ datadog, ~~falco~~)
  

use the new cluster from infra.ci.jenkins.io
	
- [x] Add the generated secrets from kubernetes in the infra.ci controller
	- [x] Add to sops
	- [x] Add to jcasc
- [x] Create a new kubernetes cloud in infra.ci
	- [x]  Manually first
	- [x]  Port as code in jcasc

migration/cleanup during this time: 

- [x] Switch the workload to the new kubernetes cloud `kubernetes-infrasicjio-agents1` 
- [x] check all jobs and for each using storage account, check if they need an extension of their network rules to add the new cluster
- [x] remove the old `kubernetes` cloud from infra.ci to avoid consuming 

### Definition of done

when the cost will be moved from the CDF payed account to the sponsored one

**Comment by smerle33** [comment]
Update: proposal for the new AKS cluster to be soon created:

    Subscription: the Azure "sponsored" subscription (same as `cijenkinsio-agents-1` https://github.com/jenkins-infra/helpdesk/issues/3954)
    Name: infracijenkinsio-agents-1. This name is valid as per [learn.microsoft.com/en-us/troubleshoot/azure/azure-kubernetes/create-upgrade-delete/aks-common-issues-faq#what-naming-restrictions-are-enforced-for-aks-resources-and-parameters-](https://learn.microsoft.com/en-us/troubleshoot/azure/azure-kubernetes/create-upgrade-delete/aks-common-issues-faq#what-naming-restrictions-are-enforced-for-aks-resources-and-parameters-)
        No dots, only hyphens, letters and number. Less than 63 characters.
        Full name of the service (infracijenkinsioagents1) to make identification easier
        agents wording to make explicit this is the only acceptable usage for this cluster
        Suffix -1 as we'll most probably need to create more clusters in the future (AWS and eventually DOKS): migration will be easier if we can increment while keeping the the same naming convention
    Connectivity:
        API access restricted to Jenkins Infra admins, privatek8s (to allow infra.ci to operate the cluster with terraform AND spawning agents)



WIP



        Should use its own subnet in the ["public_sponsorship"](https://github.com/jenkins-infra/azure-net/blob/acd7c389747988079289132236380c406a9fa119/vnets.tf#L107) virtual network.
            ⚠️ Need to refine the network sizing: This vnet already is a /14 and already has 2 x /24 subnets. Need to carefully plan the sizing of the new subnet with the AKS network rules + sizing of Nodes and pods.
            No ingress controller to be added:
                None was present on cik8s
                The ACP instance will be internal only: we'll set-it up with the Kubernetes Service internal DNS.
                    💡 If ACP cannot be used with internal SVC hostname, we can always install a "private" ingress controller as fallback, but only on last resort
            No network segregation between node pools: this cluster is not multi tenant
    Node pools:
        All Linux node pools should be Azure Linux as base OS (already done for infra.ci and release.ci controllers)
        Naming convention: given the constraint when naming node pool (ref. ):
            Linux Node pools (12 char. max.) will have:
                the OS on 1-char (l for Azure Linux, w for Windows, u for Ubuntu Linux)
                the CPU arch on 3 chars (x86 for Intel/AMD x86_64 or a64 for arm64)
                The OS on 3 chars (lin)
                The number of pods agents expected to be run on a single node of this pool (integer) preceded by a n (n3, n24, etc.) on 3 chars max
                An eventual suffix on 3 letters to specify a custom usage. May replace the sizing if needed.
            Linux node pools naming examples:
                lx86n3 => Azure Linux x86_64 nodes used which can run 3 "normal" pod agents at the same time
                lx86n4bom => Azure Linux x86_64 nodes which can run 4 ("bom" only) pod agents at the same time
                ua64n24bom => Ubuntu Linux arm64 nodes which can run 24 ("bom" only) pod agents at the same time
                la64n2side => Linux arm64 nodes used to run 2 "side" pod (e.g. custom applications such as ACP).
            Windows Node pools (6 char. max. which is trickier) will have:
                The OS (Windows) as 1-char prefix (w of course)
                The Windows edition on 4-chars (2019, 2022, etc.)
                An eventual suffix to nodepools rotation
        Expecting the following mappings (comparing to the existing cik8s EKS cluster):
            [tiny_ondemand_linux](https://github.com/jenkins-infra/aws/blob/8d0246eff56f691a34ad6035a9e0e545431cff35/cik8s-cluster.tf#L88C5-L88C24) => will be a syspool following AKS good practises (HA, etc.). Should only hosts the Azure or AKS technical side-services, not ours (CSI, CNI, etc.)
            [default_linux_az1](https://github.com/jenkins-infra/aws/blob/8d0246eff56f691a34ad6035a9e0e545431cff35/eks-public-cluster.tf#L89) => la64n2app (2 "app" pods per node: ACP and datadog-cluster's agent)
                May change to la64n3app if we add falco or any other tool
                ⚠️ Need to refine the nodes sizing
            [spot_linux_4xlarge](https://github.com/jenkins-infra/aws/blob/8d0246eff56f691a34ad6035a9e0e545431cff35/cik8s-cluster.tf#L106) => lx86n3agt1 (Azure Linux node pool for agents number "1" supporting 3x pod agents)
                ⚠️ Need to refine the nodes sizing: should be 16 vCPUS / 32 Gb / 90gb+ disk to map to current EKS sizing
                No spot as we don't have access in the Azure subscription
                Auto-scaled with minimum of 0 and maximum of 50 (same as EKS)
            [spot_linux_4xlarge_bom](https://github.com/jenkins-infra/aws/blob/8d0246eff56f691a34ad6035a9e0e545431cff35/cik8s-cluster.tf#L146) => lx86n3bom1 (Azure Linux node pool number "1" for BOM only supporting 3x pod agents)
                Same as spot_linux_4xlarge except taints to be added to ensure only bom builds are using this node pool
            [spot_linux_24xlarge_bom](https://github.com/jenkins-infra/aws/blob/8d0246eff56f691a34ad6035a9e0e545431cff35/cik8s-cluster.tf#L191) not retained

**Comment by smerle33** [comment]
the only expected "application" is the Datadog's cluster-agent (2 pods). We should reuse the system pool to run it as it's not an heavy consumer: confirmed by checking the load on the ci.jenkins.io-agents-1 cluster with the same kind of node pool.
TODO: add taint toleration to allow datadog cluster-agent to spawn on the system pool `"CriticalAddonsOnly=true:NoSchedule" taint to the default node pool`

**Comment by smerle33** [comment]
service account creation https://github.com/jenkins-infra/azure/pull/722 needed to resync manually through azure UI the peering as the subnet have been extended  see above:
⚠️ Need to refine the network sizing: This vnet already is a /23 and already has 2 x /24 subnets. Need to carefully plan the sizing of the new subnet with the AKS network rules + sizing of Nodes and pods.

**Comment by smerle33** [comment]
datadog namespace created manually:
```
>kubectl get ns                  
NAME              STATUS   AGE
datadog           Active   5s
```

**Comment by dduportal** [comment]
Reopening: The cluster is not used by infra.ci yet

**Comment by smerle33** [comment]
We had to : 
- change the certificate for the cluster to be non encoded as base64 (vs the documentation of the kubernetes plugin) - https://github.com/jenkins-infra/kubernetes-management/pull/5332
- allow access to the azure vault through the azure UI and https://github.com/jenkins-infra/azure-net/pull/258
- allow peering between this new cluster `infracijenkinsioagent1` and privatek8s and publick8s - https://github.com/jenkins-infra/azure-net/pull/259 and https://github.com/jenkins-infra/azure-net/pull/260

**Comment by dduportal** [comment]
Update: infra.ci.jenkins.io is now using the new cluster.

- Config updated by https://github.com/jenkins-infra/kubernetes-management/pull/5342 (removed the `jnlp-windows` and renamed the `jnlp-linux`)
- As the outbound IP changed, some fixes had to be done. The Terraform project Azure was "locked" due to these blocker: I had to run some hotfixes from local machine to unblock: lesson learned for later
  - Missing PR between new cluster and DB vnet: https://github.com/jenkins-infra/azure-net/pull/262
  - Missing cluster subnet in allow-list regarding IP restrictions: https://github.com/jenkins-infra/azure/commit/4ba0615085cabd89bdb4eaf54204f87fe43e75b0 and (private) https://github.com/jenkins-infra/terraform-states/commit/9f21cca289743ce465ee610d36c8be26662ed463


=> All Kubernetes, Terraform and Website jobs are green

**Comment by dduportal** [comment]
Keeping the issue opened for the week end. Some cleanup might be needed, but CDF is not paying anymore for these agents!

**Comment by dduportal** [comment]
Update: a bit of cleanup with https://github.com/jenkins-infra/azure/pull/747 and https://github.com/jenkins-infra/kubernetes-management/pull/5343

**Comment by dduportal** [comment]
Closing as it works as expected
